### ISSUE TITLE
Selection backgrounds are clobbered by rulers.

### ISSUE THREAD

**Issue by 7ombie** [helix-editor/helix#7371]
### Summary

Whenever a selection crosses over a ruler, the foreground color gets updated correctly, but the background color remains the same (the ruler color).

### Reproduction Steps

Enable rulers (adding something like `rulers = [80]` in the `[editor]` section of your `config.toml`). Then, create a selection that spans a ruler.

The background color of the ruler-column does not change to the selection color (while the foreground color updates correctly). This includes the cursor, so when you only have a cursor (a single-character selection), your cursor seems to disappear behind the ruler.

***Note: I also reported that my cursor never blinks in Helix (this was in a separate issue that somebody else opened). That issue may be contributing to the issue I'm having here.***

### Helix log

n/a

### Platform

macOS Ventura 13.4

### Terminal Emulator

iTerm2 3.4.19

### Helix Version

helix 23.05 (7f5940be)

**Comment by kirawi** [comment]
Seems like a duplicate of https://github.com/helix-editor/helix/issues/5675?

**Comment by 7ombie** [comment]
I'm not sure it's the same. Maybe, but I don't think so. This seems to be related to an oversight with the rulers specifically, which take precedence over the theme's background color (as they should), but also take precedence over the selection/cursor color (incorrectly).

#5675 changes the cursor color over whitespace, while this issue fails to adopt the selection color when the selection spans a ruler. I'll try and upload a screenshot (I'm on holiday, in a caravan).

**Comment by 7ombie** [comment]
<img width="547" alt="Screenshot 2023-06-18 at 20 39 08" src="https://github.com/helix-editor/helix/assets/80077603/08ab804b-3e8c-4c6d-a084-58f48aec86f2">

<img width="547" alt="Screenshot 2023-06-18 at 20 38 33" src="https://github.com/helix-editor/helix/assets/80077603/7c028a46-b6af-485a-b600-3c859d287015">

**Comment by gabydd** [comment]
yeah this is the same problem as in #5675 it happens because the selection(and the cursor) is rendered before the rulers, see how rulers is called after selections https://github.com/helix-editor/helix/blob/d5af6031f6ccd0a3f32efbb210aec24cd9c71ab1/helix-term/src/ui/editor.rs#L148-L154 https://github.com/helix-editor/helix/blob/d5af6031f6ccd0a3f32efbb210aec24cd9c71ab1/helix-term/src/ui/editor.rs#L199

**Comment by pascalkuthe** [comment]
We could render the rulers before the text to fix this line we do for the cursorsline/cursorcolumn. That would mean that themes that define background color for their text would clobber the rules but we are sort of leaning towards forbidding that anyways since the same caveats apply to any text with `bg` color defined.

**Comment by the-mikedavis** [comment]
There was https://github.com/helix-editor/helix/pull/3599 along those lines but foreground rulers would completely disappear. I think we could render the background and modifier parts of the ruler Style before the text highlights and then the foreground part afterwards so it would layer correctly.

**Comment by the-mikedavis** [comment]
Or I suppose we would want the background part of the ruler to be rendered on top of the text highlights but below the selection highlights.

I wonder if we should also not allow the reversed modifier to affect rulers and other decorations like indent guides? I think we could swap the foreground/backgrounds for those Styles when the existing Style has the reversed modifier.

**Comment by 7ombie** [comment]
Thanks for looking into this. Much appreciated.

**Comment by lizclipse** [comment]
This looks like it's still an issue on many themes, would a PR with [this fix](https://github.com/nilium/helix/commit/d7fe603a9a880debb165a87921d700f55a69c9b0) in be accepted? I've had a custom branch with it in for a while now and it's been working a treat, and double checking that change on current master seems to show that all themes work well with it.

**Comment by the-mikedavis** [comment]
Yeah I think that patch makes sense, I've pulled it in as 2d3b75a8c591739d7541c05ff5c6b6e3dc63d012. Originally I was apprehensive of a change like that since it could make foreground rulers hard to read. Since my last comment though we've been moving towards all themes using background rulers - foreground rulers are not really readable to begin with and should probably not be used in general.

**Comment by nilium** [comment]
I think this means I won't have to maintain a patch set for my personal install soon anymore. Thanks, @lizclipse and @the-mikedavis.

**Comment by warmwaffles** [comment]
@the-mikedavis and @nilium I really appreciate the change in 2d3b75a8c591739d7541c05ff5c6b6e3dc63d012, there is some weird behavior and I don't know if it was intended when the next character is a newline.

![Image](https://github.com/user-attachments/assets/32b86ae8-b767-4ec8-8059-7ee8fdf653de)

The next character after `longer` is `\n`. I would expect the ruler to still be rendered for that case.

Again, thank you for turning that ruler off when characters are present!

**Comment by nilium** [comment]
@warmwaffles Does this show up on a particular theme? I don't currently see it happening with mine or my build (stable release plus patch, so grain of salt on how current it is), so that seems odd. I do notice it if I have `ui.cursorline.bg` set to something. In that case, the cursorline background is drawn over the ruler, which is at least what I wanted when I made the patch.

Quick screenshot of how it looks with a newline landing on the ruler for me (and cursor below it replacing the ruler with the cursorline):

![Image](https://github.com/user-attachments/assets/6ffad71a-3088-454d-8ef5-48c779de07ac)

Probably worth opening a new issue just in case anyway.

**Comment by warmwaffles** [comment]
I am using https://github.com/folke/tokyonight.nvim/blob/main/extras/helix/tokyonight_moon.toml

Let me swap to a new theme, maybe it's a problem upstream.
